### PR TITLE
Allow specification of namespaces to use in RemapJarTask

### PR DESCRIPTION
**Pull request by i509VCB**
This is a draft since setting `toM` as `official` has mapping conflicts, this is probably out of our control tbh.

But for the future this may have additional use for exporting to other mapping sets (such as mojmap) in the future.

### CODE REVIEW

File: src/main/java/net/fabricmc/loom/task/RemapJarTask.java [hunk 1]
```diff
@@ -176,8 +197,20 @@ public void scheduleRemap() throws Throwable {
...
+		List<String> namespaces = mappings.getMetadata().getNamespaces();
+
+		// Validate the namespaces are supported in the tiny tree
+		if (!namespaces.contains(fromM)) {
+			RemapJarTask.throwMissingNamespaceException(project, fromM, namespaces);
+		}
+
+		if (!namespaces.contains(toM)) {
+			RemapJarTask.throwMissingNamespaceException(project, toM, namespaces);
+		}
```
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/task/RemapJarTask.java]:
```suggestion
		if (!namespaces.contains(fromM)) {
			throwMissingNamespaceException(project, fromM, namespaces);
		}

		if (!namespaces.contains(toM)) {
			throwMissingNamespaceException(project, toM, namespaces);
		}
```
There's no need to specify the class name here.

File: src/main/java/net/fabricmc/loom/task/RemapJarTask.java [hunk 2]
```diff
@@ -91,8 +100,20 @@ public void doSingleRemap() throws Throwable {
...
+		List<String> namespaces = mappings.getMetadata().getNamespaces();
+
+		// Validate the namespaces are supported in the tiny tree the namespaces are supported in the tiny tree
+		if (!namespaces.contains(fromM)) {
+			RemapJarTask.throwMissingNamespaceException(project, fromM, namespaces);
+		}
+
+		if (!namespaces.contains(toM)) {
+			RemapJarTask.throwMissingNamespaceException(project, toM, namespaces);
+		}
```
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/task/RemapJarTask.java]:
```suggestion
		if (!namespaces.contains(fromM)) {
			throwMissingNamespaceException(project, fromM, namespaces);
		}

		if (!namespaces.contains(toM)) {
			throwMissingNamespaceException(project, toM, namespaces);
		}
```
Same here.

**Review by Juuxel — CHANGES_REQUESTED** [review]
Looks nice and this would be useful in one of my Loom fork's branches :eyes:

File: src/main/java/net/fabricmc/loom/task/RemapJarTask.java (hunk 1 above)
Comment (i509VCB) [reply src/main/java/net/fabricmc/loom/task/RemapJarTask.java]: I qualify these things so I know whether a method is static or belongs to a class file (usually via `this`)

Comment (i509VCB) [reply src/main/java/net/fabricmc/loom/task/RemapJarTask.java]: Of course no conventions exist on whether member references should be qualified or not

**Comment by LeafHacker** [comment]
This looks like it would resolve #191, could you clarify what you mean by _“official has mapping conflicts, this is probably out of our control tbh”_? Are you saying Tiny remapper has issues mapping a jar to `official` because the member names are sometimes duplicated?

**Comment by modmuss50** [comment]
This is also missing changing the output of the AW remapping, and the mixin refmap + output mappings

**Comment by modmuss50** [comment]
This was handled when I rewrote this. Thanks.

**Pull request closed without merging**